src/dlr_tvm.cc

```diff
@@ -1,105 +1,100 @@
 #include "dlr_tvm.h"
 
 #include <stdlib.h>
+
 #include <fstream>
 #include <iterator>
 #include <numeric>
 
 using namespace dlr;
 
-ModelPath dlr::GetTvmPaths(std::vector<std::string> dirname) {
-  ModelPath paths;
-  std::vector<std::string> paths_vec;
-  for (auto dir : dirname) {
-    ListDir(dir, paths_vec);
-  }
-  for (auto filename : paths_vec) {
+void TVMModel::InitModelArtifact() {
+  std::shared_ptr<TVMModelArtifact> artifact =
+      std::make_shared<TVMModelArtifact>();
+  std::vector<std::string> filenames = ListFilesInDirectories(paths_);
+  for (auto filename : filenames) {
     std::string basename = GetBasename(filename);
     if (EndsWith(filename, ".json") &&
         std::all_of(
             std::begin(SAGEMAKER_AUXILIARY_JSON_FILES),
             std::end(SAGEMAKER_AUXILIARY_JSON_FILES),
             [basename](const std::string& s) { return (s != basename); }) &&
         filename != "version.json") {
-      paths.model_json = filename;
+      artifact->model_json = filename;
     } else if (filename != LIBDLR && EndsWith(filename, LIBEXT)) {
-      paths.model_lib = filename;
+      artifact->model_lib = filename;
     } else if (EndsWith(filename, ".tensorrt")) {
-      paths.model_lib = filename;
+      artifact->model_lib = filename;
     } else if (EndsWith(filename, ".params")) {
-      paths.params = filename;
+      artifact->params = filename;
     } else if (filename == "version.json") {
-      paths.ver_json = filename;
+      artifact->ver_json = filename;
     } else if (EndsWith(filename, ".meta")) {
-      paths.metadata = filename;
+      artifact->metadata = filename;
     }
   }
-  if (paths.model_json.empty() || paths.model_lib.empty() ||
-      paths.params.empty()) {
+  if (artifact->model_json.empty() || artifact->model_lib.empty() ||
+      artifact->params.empty()) {
     LOG(INFO) << "No valid TVM model files found under folder:";
-    for (auto dir : dirname) {
+    for (auto dir : paths_) {
       LOG(INFO) << dir;
     }
     LOG(FATAL);
+  } else {
+    model_artifact_ = std::shared_ptr<ModelArtifact>(artifact);
```

Comment:
Use ` model_artifact_ = std::move(artifact);`

This can be done for Hexagon and treelite too probably.

https://stackoverflow.com/questions/41871115/why-would-i-stdmove-an-stdshared-ptr

Reply:
:+1:

Reply:
I think unique_ptr is better in that case.